server/src/main/java/org/opensearch/gateway/ShardsBatchGatewayAllocator.java

```diff
@@ -0,0 +1,734 @@
+    ) {
+        throw new UnsupportedOperationException("ShardsBatchGatewayAllocator does not support allocating unassigned shards");
+    }
+
+    @Override
+    public void allocateAllUnassignedShards(final RoutingAllocation allocation, boolean primary) {
+
+        assert primaryShardBatchAllocator != null;
+        assert replicaShardBatchAllocator != null;
+        innerAllocateUnassignedBatch(allocation, primaryShardBatchAllocator, replicaShardBatchAllocator, primary);
+    }
+
+    protected void innerAllocateUnassignedBatch(
+        RoutingAllocation allocation,
+        PrimaryShardBatchAllocator primaryBatchShardAllocator,
+        ReplicaShardBatchAllocator replicaBatchShardAllocator,
+        boolean primary
+    ) {
+        // create batches for unassigned shards
+        Set<String> batchesToAssign = createAndUpdateBatches(allocation, primary);
+        if (batchesToAssign.isEmpty()) {
+            return;
+        }
+        if (primary) {
+            batchIdToStartedShardBatch.values()
+                .stream()
+                .filter(batch -> batchesToAssign.contains(batch.batchId))
+                .forEach(
+                    shardsBatch -> primaryBatchShardAllocator.allocateUnassignedBatch(shardsBatch.getBatchedShardRoutings(), allocation)
+                );
+        } else {
+            batchIdToStoreShardBatch.values()
+                .stream()
+                .filter(batch -> batchesToAssign.contains(batch.batchId))
+                .forEach(batch -> replicaBatchShardAllocator.allocateUnassignedBatch(batch.getBatchedShardRoutings(), allocation));
+        }
+    }
+
+    // visible for testing
+    protected Set<String> createAndUpdateBatches(RoutingAllocation allocation, boolean primary) {
+        Set<String> batchesToBeAssigned = new HashSet<>();
+        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
+        ConcurrentMap<String, ShardsBatch> currentBatches = primary ? batchIdToStartedShardBatch : batchIdToStoreShardBatch;
+        // get all batched shards
+        Set<ShardId> currentBatchedShards = currentBatches.values()
+            .stream()
+            .map(ShardsBatch::getBatchedShards)
+            .flatMap(Set::stream)
+            .collect(Collectors.toSet());
+
+        Set<ShardRouting> shardsToBatch = Sets.newHashSet();
+        // add all unassigned shards to the batch if they are not already in a batch
+        unassigned.forEach(shardRouting -> {
+            if ((currentBatchedShards.contains(shardRouting.shardId()) == false) && (shardRouting.primary() == primary)) {
+                assert shardRouting.unassigned();
+                shardsToBatch.add(shardRouting);
+            }
+            // if shard is already batched update to latest shardRouting information in the batches
+            else if (shardRouting.primary() == primary) {
+                String batchId = getBatchId(shardRouting, shardRouting.primary());
```

Comment:
it is iterating over all shards in all batches to find the batch and this will run for every unassigned for fetching is already in progress and a reroute is triggered. Worst case this could be O(n2) where n is no. of unassigned shards. Instead of creating set of currentBatchedShards at line no. 229, you create map for shardId -> Batch lookup which is internal to this method and then you can use it both to check whether shard is already batched and to which batch

Reply:
Ack.